rest-server/src/models/job.js

```diff
@@ -41,42 +41,61 @@ class Job {
     });
   }
 
+  convertJobState(frameworkState, exitCode) {
+    let jobState = '';
+    switch(frameworkState) {
+      case 'FRAMEWORK_WAITING':
+      case 'APPLICATION_CREATED':
+      case 'APPLICATION_LAUNCHED':
+      case 'APPLICATION_WAITING':
+        jobState = 'WAITING';
+        break;
+      case 'APPLICATION_RUNNING':
+      case 'APPLICATION_RETRIEVING_DIAGNOSTICS':
+      case 'APPLICATION_COMPLETED':
+        jobState = 'RUNNING';
+        break;
+      case 'FRAMEWORK_COMPLETED':
+        if (exitCode && parseInt(exitCode) === 0) {
+          jobState = 'SUCCEEDED';
+        } else {
+          jobState = 'FAILED';
+        }
+        break;
+      default:
+        jobState = 'UNKNOWN';
+    }
+    return jobState;
+  }
+
   getJobList(next) {
     unirest.get(launcherConfig.frameworksPath())
         .headers(launcherConfig.webserviceRequestHeaders)
         .end((res) => {
           const resJson = typeof res.body === 'object' ?
               res.body : JSON.parse(res.body);
-          const jobList = [];
-          async.each(resJson.frameworkNames, (frameworkName, callback) => {
-            this.getJobStatus(frameworkName, (jobStatus) => {
-              const jobOverview = {
-                name: frameworkName,
-                username: jobStatus.username,
-                state: jobStatus.state,
-                subState: jobStatus.subState,
-                retries: jobStatus.retries,
-                createdTime: jobStatus.createdTime,
-                completedTime: jobStatus.completedTime,
-                appTrackingUrl: jobStatus.appTrackingUrl,
-                appExitType: jobStatus.appExitType
-              };
-              jobList.push(jobOverview);
-              callback();
+          const jobList = res.summarizedFrameworkInfos.map((frameworkInfo) => {
+            let retries = 0;
+            ['transientNormalRetriedCount', 'transientConflictRetriedCount',
+                'nonTransientRetriedCount', 'unKnownRetriedCount'].forEach((retry) => {
+              retries += frameworkInfo.frameworkRetryPolicyState[retry];
             });
-          }, (err) => {
-            jobList.sort((a, b) => b.createdTime - a.createdTime);
-            next(jobList, err);
+            return {
+              name: frameworkInfo.frameworkName,
+              username: frameworkInfo.userName,
+              state: this.convertJobState(frameworkInfo.frameworkState, frameworkInfo.applicationExitCode),
+              subState: frameworkInfo.frameworkState,
+              retries: retries,
+              createdTime: frameworkInfo.firstRequestTimestamp || new Date(2018, 1, 1).getTime(),
```

Comment:
In what circumstance will this field be empty? And I think we shall not use a default value.

Reply:
The old frameworks do not have this field, we ensure new frameworks will always have.

---
In reply to: [167141691](https://github.com/Microsoft/pai/pull/314#discussion_r167141691) [](ancestors = 167141691)